### PR TITLE
Eth to DAI LBP

### PR DESCRIPTION
**Pull request by Joeysantoro**
# PR Type: Feature

### PR Description: This proposal accomplishes two ETH PCV related actions.
1. An LBP of 20000 ETH to DAI
2. Tighten Eth PSM redemption fee from 75bps to 60bps

### PR Checklist - Feature (Proposal)

- [x] All Tests Passing
- [x] Proposal Added to ProposalsConfig
- [x] Fork Block Correct
- [x] Remove Any .only's on Tests
- [x] Update Documentation If Needed
- [x] Update Roles Config
- [ ] Proposal Submitted

### CODE REVIEW

**Review by xklob — DISMISSED** [review]

File: proposals/dao/eth_lbp.ts [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+  // Validate PSM_ADMIN_ROLE is under ROLE_ADMIN and that TribalCouncilTimelock has the role
+  expect(await core.hasRole(ethers.utils.id('PSM_ADMIN_ROLE'), addresses.tribalCouncilTimelock)).to.be.true;
+  expect(await core.getRoleAdmin(ethers.utils.id('PSM_ADMIN_ROLE'))).to.be.equal(ethers.utils.id('ROLE_ADMIN'));
+};
+
+const validateLBPSetup = async (contracts: NamedContracts, addresses: NamedAddresses, poolId: string) => {
+  const ethToDaiLBPSwapper = contracts.ethToDaiLBPSwapper;
+
+  // expect(await ethToDaiLBPSwapper.doInvert()).to.be.equal(true);
```
Comment (thomas-waite) [inline proposals/dao/eth_lbp.ts]: Would probably uncomment and validate it's false

File: proposals/dao/eth_lbp.ts [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+const LBP_FREQUENCY = 86400 * 2; // 2 days in seconds
+const MIN_LBP_SIZE = ethers.constants.WeiPerEther.mul(500); // 500 ETH
+let poolId; // auction pool id
+
+const fipNumber = '110';
+
+const deploy: DeployUpgradeFunc = async (deployAddress: string, addresses: NamedAddresses, logging: boolean) => {
+  ///////////  2. Deploy the Balancer LBP swapper
+  // // Amounts:
+  // WETH: 37888449801955370645659 (95%), 37k WETH, $3,587,445
```
Comment (thomas-waite) [inline proposals/dao/eth_lbp.ts]: Would probably update these comments to be accurate, no strong feelings though

File: proposals/description/eth_lbp.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      values: '0',
+      method: 'withdrawToSafeAddress(address,address,uint256,bool,bool)',
+      arguments: ['{compoundDaiPCVDeposit}', '{ethToDaiLBPSwapper}', '3000000000000000000000000', false, false],
+      description: 'Withdraw 3M DAI from the CompoundPCVDeposit and transfer to the LBP pool'
+    },
+    // Swap and update CR oracle
+    {
+      target: 'ethToDaiLBPSwapper',
+      values: '0',
+      method: 'swap()',
```
Comment (thomas-waite) [inline proposals/description/eth_lbp.ts:32]: Makes sense to use `swap()`. Duration is 2 days, and TC timelock takes 4 days. `forceSwap()` is `onlyGovernor`

**Review by thomas-waite — COMMENTED** [review]
Looks good to me, would clear up some of the comments

File: proposals/description/eth_lbp.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import { ProposalDescription } from '@custom-types/types';
+
+const eth_lbp: ProposalDescription = {
+  title: 'FIP-110: Reinforce PCV via ETH LBP and tighter spread',
+  commands: [
+    {
+      target: 'pcvGuardianNew',
+      values: '0',
+      method: 'setSafeAddresses(address[])',
```
Comment (xklob) [inline proposals/description/eth_lbp.ts:9]: Metacomment - there is a singular version of this function that is easier to call

**Review by xklob — DISMISSED** [review]
Looks good.

File: proposals/dao/eth_lbp.ts [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+  const daiSpent = initialUserDaiBalance.sub(postUserDaiBalance);
+  expect(daiSpent).to.be.bignumber.equal(amountIn);
+
+  const ethGained = postUserEthBalance.sub(initialUserEthBalance);
+  expect(ethGained).to.be.bignumber.at.least(ethers.constants.WeiPerEther.mul(4));
+  expect(ethGained).to.be.bignumber.at.most(ethers.constants.WeiPerEther.mul(6));
+
+  // Put in 10k DAI, got out ~5 WETH
+  // Implies price of $2000 per WETH
```
Comment (thomas-waite) [inline proposals/dao/eth_lbp.ts:276]: Nice

**Review by thomas-waite — COMMENTED** [review]
Looks good to me! Want to deploy and verify the contracts, then add to this PR before merging?

**Comment by Joeysantoro** [comment]
Contracts deployed and verified

Calldata: 
```
0x8f2a0bb000000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000001c000000000000000000000000000000000000000000000000000000000000002c000000000000000000000000000000000000000000000000000000000000000000d0c2b4e41c5fdf9375f2b1601caf0af18e81d6257ea3abb3b72a989fc3698da0000000000000000000000000000000000000000000000000000000000054600000000000000000000000000000000000000000000000000000000000000000700000000000000000000000002435948f84d7465fb71de45aba6098fc6ec299300000000000000000000000002435948f84d7465fb71de45aba6098fc6ec299300000000000000000000000002435948f84d7465fb71de45aba6098fc6ec2993000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a54000000000000000000000000ff6f59333cfd8f4ebc14ad0a0e181a83e655d257000000000000000000000000ff6f59333cfd8f4ebc14ad0a0e181a83e655d25700000000000000000000000098e5f5706897074a4664dd3a32eb80242d6e694b00000000000000000000000000000000000000000000000000000000000000070000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000700000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000002600000000000000000000000000000000000000000000000000000000000000340000000000000000000000000000000000000000000000000000000000000038000000000000000000000000000000000000000000000000000000000000003e0000000000000000000000000000000000000000000000000000000000000044000000000000000000000000000000000000000000000000000000000000000648320357d00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a540000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a4fe68d76e0000000000000000000000005b86887e171bae0c2c826e87e34df8d558c079b9000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a5400000000000000000000000000000000000000000000043c33c1937564800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a4fe68d76e000000000000000000000000e0f73b8d76d2ad33492f995af218b03564b8ce20000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a54000000000000000000000000000000000000000000027b46536c66c8e3000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000048119c065000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000024dea02892000000000000000000000000df9ff5c077d9f3427ade67ac2d27a864be6f3187000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000024dea02892000000000000000000000000f24401f6992faeacbc5d6c6991db15b5f8364a1b0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000245d841af5000000000000000000000000000000000000000000000000000000000000003c00000000000000000000000000000000000000000000000000000000
```

Execute Calldata: 
```
0xe38335e500000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000001a000000000000000000000000000000000000000000000000000000000000002a000000000000000000000000000000000000000000000000000000000000000000d0c2b4e41c5fdf9375f2b1601caf0af18e81d6257ea3abb3b72a989fc3698da000000000000000000000000000000000000000000000000000000000000000700000000000000000000000002435948f84d7465fb71de45aba6098fc6ec299300000000000000000000000002435948f84d7465fb71de45aba6098fc6ec299300000000000000000000000002435948f84d7465fb71de45aba6098fc6ec2993000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a54000000000000000000000000ff6f59333cfd8f4ebc14ad0a0e181a83e655d257000000000000000000000000ff6f59333cfd8f4ebc14ad0a0e181a83e655d25700000000000000000000000098e5f5706897074a4664dd3a32eb80242d6e694b00000000000000000000000000000000000000000000000000000000000000070000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000700000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000002600000000000000000000000000000000000000000000000000000000000000340000000000000000000000000000000000000000000000000000000000000038000000000000000000000000000000000000000000000000000000000000003e0000000000000000000000000000000000000000000000000000000000000044000000000000000000000000000000000000000000000000000000000000000648320357d00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a540000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a4fe68d76e0000000000000000000000005b86887e171bae0c2c826e87e34df8d558c079b9000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a5400000000000000000000000000000000000000000000043c33c1937564800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a4fe68d76e000000000000000000000000e0f73b8d76d2ad33492f995af218b03564b8ce20000000000000000000000000f7991f4698ffb6716982aec7f78964dd731c4a54000000000000000000000000000000000000000000027b46536c66c8e3000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000048119c065000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000024dea02892000000000000000000000000df9ff5c077d9f3427ade67ac2d27a864be6f3187000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000024dea02892000000000000000000000000f24401f6992faeacbc5d6c6991db15b5f8364a1b0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000245d841af5000000000000000000000000000000000000000000000000000000000000003c00000000000000000000000000000000000000000000000000000000
```

File: proposals/dao/eth_lbp.ts [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+  console.log('DAI spent: ', daiSpent);
+  console.log('WETH gained: ', ethGained);
+
+  // Accelerate time and check ended
+  await time.increase(LBP_FREQUENCY);
```
Comment (thomas-waite) [inline proposals/dao/eth_lbp.ts:281]: I also fast forwarded time, performed a second swap and validated that the amountOut increased - to check that the weights were decreasing as expected. No strong feelings, it's a bit paranoid

File: protocol-configuration/mainnetAddresses.ts [hunk 7]
```diff
@@ -196,6 +196,26 @@ const MainnetAddresses: MainnetAddresses = {
...
+    category: AddressCategory.PCV
+  },
+  ethToDaiLensEth: {
+    artifactName: 'BPTLens',
+    address: '0xf24401f6992faeacbc5d6c6991db15b5f8364a1b',
+    category: AddressCategory.PCV
+  },
+  ethToDaiLBPPool: {
+    artifactName: 'IWeightedPool',
+    address: '0x34809aEDF93066b49F638562c42A9751eDb36DF5',
```
Comment (thomas-waite) [inline protocol-configuration/mainnetAddresses.ts:216]: Would recommend verifying this LBPPool contract and inspecting the variables. Neat trick to verify these LBP pools is to pull the source code of an already verified contract and then use the standard-json input verification method:

```
curl -s [https://api.etherscan.io/api\?module\=contract\&action\=getsourcecode\&address\=0x32F0DC9E2D890BAC95106a65eB82db70Bc58BaDB\&apikey\=AETHERSCANPIKEY](https://api.etherscan.io/api/?module\=contract\&action\=getsourcecode\&address\=0x32F0DC9E2D890BAC95106a65eB82db70Bc58BaDB\&apikey\=AETHERSCANPIKEY) | jq -r '.result[0].SourceCode' | sed 's/{{/{/' | sed 's/}}/}/' > sol.json
```

**Review by thomas-waite — DISMISSED** [review]
Looks good to me, minor comments

**Review by thomas-waite — APPROVED** [review]

**Pull request merged**